### PR TITLE
Feature/120 display time

### LINKED ISSUE #120: Option to show time since start of the release
# Situation

Waiting for a release to finish can make a user nervous. So it would be nice to have a kind of a stop-watch running that shows and updates the time since the release was started.

Updated text is only useful in interactive terminals, so it should be off by default.

## Acceptance Criteria

* Switch added to display time since release start
* When activated, the timer should be displayed and updated each second
* Optional: if possible it would be nice if the time of the last release build could be displayed as a reference (depending on whether the GH API supports that)
* Additional proposal by @kaklakariada: release-droid displays the link of the current workflow run in github to enable the user to visit the script directly

### PR DESCRIPTION
**Pull request by ckunki**
fixes #120

### CODE REVIEW

File: src/main/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatter.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+            return String.format("~ %d minute%s", m, plural(m));
+        }
+
+        final long s = duration.abs().toSeconds();
+        return String.format("%d second%s", s, plural(s));
+    }
+
+    static LocalDateTime localDateTime(final Date date) {
+        return LocalDateTime.ofInstant(date.toInstant(), ZoneOffset.UTC);
+    }
```
Comment (kaklakariada) [inline src/main/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatter.java:159]: Use Instant as long as possible, convert to LocalDateTime for output:
```java
LocalDateTime.ofInstant(instant, ZoneId.systemDefault())
```

File: src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+            "0:00:44, 44 seconds" //
+    })
+    void formatRemaining(final String remaining, final String expected) {
+        final Duration duration = Duration.parse(String.format("PT%sH%sM%sS", (Object[]) remaining.split(":")));
+        assertThat(ProgressFormatter.formatRemaining(duration), equalTo(expected));
+    }
+
+    @Test
+    void welcomeMessage() throws InterruptedException {
+        final Duration duration = Duration.ofHours(1).plusMinutes(2);
```
Comment (kaklakariada) [inline src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java:61]: Use DURATION instead

File: src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                .thenReturn(Duration.ofSeconds(59));
+        final ProgressFormatter testee = startFormatter(monitor, estimation);
+
+        final String status1 = testee.status();
+        assertThat(status1, containsString(green("0:00:00 elapsed")));
+        assertThat(status1, containsString(yellow("~ 1 minute remaining")));
+        // sleep 1 s
+        final String status2 = testee.status();
+        assertThat(status2, containsString(green("0:00:01 elapsed")));
+        assertThat(status2, containsString(yellow("59 seconds remaining")));
```
Comment (kaklakariada) [inline src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java:103]: Use matcher  `allOf()` to combine assertions

e.g.
```
assertThat(status2, allOf(containsString(yellow("59 seconds remaining")),containsString(green("0:00:01 elapsed"))));
```

File: src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+                { "0:00:02 elapsed", "1 second overdue", "200%", "[=========", "|=========>" } };
+        for (int i = 0; i < 3; i++) {
+            verifyContents(testee.status(), expected[i]);
+        }
+    }
+
+    private void verifyContents(final String actual, final String... expected) {
+        for (final String e : expected) {
+            assertThat(actual, containsString(e));
+        }
```
Comment (kaklakariada) [inline src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java:154]: Proposal: convert list of expected values to array of matchers, then use `allOf(matchers)`

File: src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    void manualIntegrationTestWithoutGithub() throws InterruptedException {
+        final Duration estimation = Duration.ofSeconds(4);
+        final ProgressFormatter testee = startFormatter(new ProgressMonitor(), estimation);
+        new ManualExplorer().estimation(estimation).iterations(5, 3).run(testee, "");
+    }
+
+    void manualIntegrationTestWithGithub() throws IOException, GitHubException, InterruptedException {
+        final String RELEASE_DROID_CREDENTIALS = RELEASE_DROID_DIRECTORY + FILE_SEPARATOR + "credentials";
```
Comment (kaklakariada) [inline src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java:164]:
```suggestion
        final String credentials = RELEASE_DROID_DIRECTORY + FILE_SEPARATOR + "credentials";
```

File: src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        final PropertyReaderImpl reader = new PropertyReaderImpl(RELEASE_DROID_CREDENTIALS);
```
Comment (kaklakariada) [inline src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java:165]:
```suggestion
        final PropertyReaderImpl reader = new PropertyReaderImpl(credentials);
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatter.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        return new Builder();
+    }
+
+    private static final String ANSI_GREEN = "\u001B[92m";
+    private static final String ANSI_RED = "\u001B[31m";
+    private static final String ANSI_YELLOW = "\u001B[93m";
+
+    private final ProgressMonitor monitor;
+    private DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HH:mm:ss");
+    private DateTimeFormatter dateFormatter = DateTimeFormatter.ISO_LOCAL_DATE;
```
[the remark this answers is not in the document]
Comment (ckunki) [reply src/main/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatter.java:22]: ProgressFormatter.Builder violates the builder pattern a little bit and sets the fields of the built instance directly ....
This is only possible if fields are non-final :(

File: src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java [hunk 8]
```diff
@@ -144,14 +142,13 @@ void progress() throws InterruptedException {
...
         }
     }
 
-    private void verifyContents(final String actual, final String... expected) {
-        for (final String e : expected) {
-            assertThat(actual, containsString(e));
-        }
+    @SuppressWarnings("unchecked")
+    private Matcher<String> matchers(final String... expected) {
+        return allOf(Arrays.stream(expected).map(Matchers::containsString).toArray(Matcher[]::new));
```
Comment (kaklakariada) [inline src/test/java/com/exasol/releasedroid/adapter/github/progress/ProgressFormatterTest.java:151]: 👍

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**